src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -6483,41 +6514,71 @@ public void Invoke(Task completingTask)
+        /// <remarks>
+        /// The returned task will complete when any of the supplied tasks has completed.  The returned task will always end in the RanToCompletion state
+        /// with its Result set to the first task to complete.  This is true even if the first task to complete ended in the Canceled or Faulted state.
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        /// The <paramref name="tasks"/> collection contained a null task, or was empty.
+        /// </exception>
+        private static Task<TTask> WhenAny<TTask>(IEnumerable<TTask> tasks) where TTask : Task
         {
             // Skip a List allocation/copy if tasks is a collection
-            if (tasks is ICollection<Task> taskCollection)
+            if (tasks is ICollection<TTask> tasksAsCollection)
             {
-                // Take a more efficient path if tasks is actually an array
-                if (tasks is Task[] taskArray)
+                // Take a more efficient path if tasks is actually a list or an array. Arrays are a bit less common,
+                // since if argument was strongly-typed as an array, it would have bound to the array-based overload.
+                if (tasks is List<TTask> tasksAsList)
                 {
-                    return WhenAny(taskArray);
+                    return WhenAny((ReadOnlySpan<TTask>)CollectionsMarshal.AsSpan(tasksAsList));
+                }
+                if (tasks is TTask[] tasksAsArray)
+                {
+                    return WhenAny((ReadOnlySpan<TTask>)tasksAsArray);
                 }
 
-                int count = taskCollection.Count;
+                int count = tasksAsCollection.Count;
                 if (count <= 0)
                 {
                     ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_EmptyTaskList, ExceptionArgument.tasks);
                 }
 
-                int index = 0;
-                taskArray = new Task[count];
-                foreach (Task task in tasks)
+                var taskArray = new TTask[count];
+                tasksAsCollection.CopyTo(taskArray, 0);
+                foreach (TTask task in taskArray)
                 {
-                    if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
-                    taskArray[index++] = task;
+                    if (task is null)
+                    {
+                        ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
+                    }
                 }
+
                 return TaskFactory.CommonCWAnyLogic(taskArray);
             }
 
-            if (tasks == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
+            if (tasks is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
+            }
```

Comment:
To confirm my understanding, this check is below the specialization branches because those branches will do the same check, so this keeps us from redundantly checking for `is null`?

Reply:
Yup